aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SyncProcessor.java

```diff
@@ -131,9 +137,38 @@ Completable hydrate() {
             });
     }
 
-    private Completable createHydrationTask(Class<? extends Model> modelClass) {
+    private Completable createHydrationTask(Class<? extends Model> modelClass,
+                                            ConcurrentLinkedQueue<String> hydratedModels) {
         ModelSyncMetricsAccumulator metricsAccumulator = new ModelSyncMetricsAccumulator(modelClass);
+
+        ModelSchema schema = modelSchemaRegistry.getModelSchemaForModelClass(
+            modelClass.getSimpleName());
+        List<String> dependencies = new ArrayList<>(
+            schema.getAssociations().values().stream()
+                .filter((i) -> i.isOwner())
+                .map((i) -> i.getAssociatedType())
+                .collect(Collectors.toList()));
+
+        LOG.debug("Sync dependencies for model:" + schema.getName() + " - "
+            + dependencies.toString());
+
         return syncTimeRegistry.lookupLastSyncTime(modelClass)
+            .delaySubscription(
+                Flowable.interval(SYNC_SUBSCRIPTION_SWITCH_MILLISECONDS, TimeUnit.MILLISECONDS)
```

Comment:
Instead of checking every 100ms if the dependencies are re-evaluated, is there a way we only re-evaluate whenever any object is added to `hydratedModels`?   That way, we aren't performing this check so often.

I'm thinking you could have a `PublishSubject`, which emits the array of hydratedModels each time a new model is added to it.  Then, each hydrationTask would be an observer on that subject, and would re-evaluate the dependencies to see if it can start the sync.

I do think this PR in its current state is still an improvement over the current implementation, so this could be a follow up PR if you'd prefer.

Reply:
I like that idea, it better leans into reactive paradigms. Will do a followup commit into this PR over the next 24 hrs